stdlib/buffer.ml

```diff
@@ -87,102 +96,91 @@ let resize b more =
      this tricky function that is slow anyway. *)
-  Bytes.blit b.buffer 0 new_buffer 0 b.position;
-  b.buffer <- new_buffer;
-  b.length <- !new_len;
-  assert (b.position + more <= b.length);
-  assert (old_pos + more <= b.length);
+  Bytes.blit b.inner.buffer 0 new_buffer 0 b.position;
+  b.inner <- { buffer = new_buffer; length = !new_len };
+  assert (b.position + more <= b.inner.length);
+  assert (old_pos + more <= b.inner.length);
   ()
-  (* Note: there are various situations (preemptive threads, signals and
-     gc finalizers) where OCaml code may be run asynchronously; in
-     particular, there may be a race with another user of [b], changing
-     its mutable fields in the middle of the [resize] call. The Buffer
-     module does not provide any correctness guarantee if that happens,
-     but we must still ensure that the datastructure invariants hold for
-     memory-safety -- as we plan to use [unsafe_{get,set}].
-
-     There are two potential allocation points in this function,
-     [ref] and [Bytes.create], but all reads and writes to the fields
-     of [b] happen before both of them or after both of them.
-
-     We therefore assume that [b.position] may change at these allocations,
-     and check that the [b.position + more <= b.length] postcondition
-     holds for both values of [b.position], before or after the function
-     is called. More precisely, the following invariants must hold if the
-     function returns correctly, in addition to the usual buffer invariants:
-     - [old(b.position) + more <= new(b.length)]
-     - [new(b.position) + more <= new(b.length)]
-     - [old(b.length) <= new(b.length)]
-
-     Note: [b.position + more <= old(b.length)] does *not*
-     hold in general, as it is precisely the case where you need
-     to call [resize] to increase [b.length].
-
-     Note: [assert] above does not mean that we know the conditions
-     always hold, but that the function may return correctly
-     only if they hold.
-
-     Note: the other functions in this module does not need
-     to be checked with this level of scrutiny, given that they
-     read/write the buffer immediately after checking that
-     [b.position + more <= b.length] hold or calling [resize].
-  *)
+
+(* Note:
+    Some of the functions below have a fast path when the inner
+  buffer doesn't need to be extended.
+    In this case, it is possible to use unsafe accesses on the
+  contents of the [inner] field since its fields are immutable.
+  In presence of data races, we may access the wrong inner buffer, but we
+  will use this buffer safely.
+  As soon as we need to resize the buffer, we fall back to safe accesses.
+*)
 
 let add_char b c =
   let pos = b.position in
-  if pos >= b.length then resize b 1;
-  Bytes.unsafe_set b.buffer pos c;
+  let {buffer;length} = b.inner in
+  if pos >= length then (
+    resize b 1;
+    Bytes.set b.inner.buffer pos c
+  ) else
+    Bytes.unsafe_set buffer pos c;
   b.position <- pos + 1
```

Comment:
My own take on this (and the other functions) would be to make the basic `resize` function return the inner buffer (with its length) instead of setting it. This way you can still use unsafe accesses, and you don't have to worry about add_char looping infinitely if run in parallel with `reset` operations. You could even share the code for updating the buffer and the position, although that will likely cause a small performance hit because of the introduction of spill/reload instructions from the cold path into the hot path.

Reply:
After a quick glance, the benefits of both the recursive and the `resize-once` variants are not obvious in term of performance:

|                               | 1st rank | 2nd rank | 3rd rank | other |
|-------------------------------|----------|----------|----------|-------|
| indirection                   | 20.2%    | 46.4%    | 19.0%    | 14.3% |
| indirection+rec               | 32.1%    | 19.0%    | 36.9%    | 11.9% |
| indirection+resize-once       | 44.0%    | 16.7%    | 35.7%    | 3.6% |
| indirection+resize-once+spill | 1.2%    | 11.9%    | 4.8%    | 82.1% |
| std                           | 2.4%    | 6.0%    | 3.6%    | 88.1% |

- `indirection`: this PR
- `indirection+rec`: @gasche proposition
- `indirection+resize-once`: @lthls proposition
- `indirection+resize-once+spill` : @lthls proposition with maximal code sharing between branches
- `std` : current stdlib implementation

(This is on my linux machine that was favorable to the `indirection` test in the earlier results)

Restraining the comparison to the fast variants of the current implementation gives me the impression that there is no variant really faster than the other:

|                               | 1st rank | 2nd rank | 3rd rank |
|-------------------------------|----------|----------|----------|
| indirection+resize-once       | 44.0%    | 16.7%    | 39.3%    | 
| indirection+rec               | 31.0%    | 29.8%    | 39.3%    | 
| indirection                   | 25.0%    | 53.6%    | 21.4%    |

Note that this is on the `medium` variant of the benchmark which might be sensitive to the noise. I will share the result of the larger variant once the run is done later this evening.

Reply:
My recursive add_char is not supposed to be faster (it should have basically the same performance), but to be simpler (to the human reader; not having a safe and unsafe instructions close to each other) and shorter (in term of generated code size, I expect a sizeable reduction). Code-size effects tend to come back to bite us from time to time (weird code-cache questions in real-world large programs, but also js_of_ocaml etc.), so it's something to keep in mind.

@lthls's suggestion could also be nice as long as it is also simpler and shorter. (The easiest/nicest way to write it is with a `let buffer = ensure_room b.inner pos 1 in` function, but I would not expect this to be performance-preserving because we either pay a function call in the hot path or inline a lot more code. The current code structure, and even more the recursive version, are good at hiding the cold path away.)

Reply:
I'm starting to regret this thread of conversation -- guilty. Why don't we just take @Octachron's code and run with it?

Reply:
@Octachron on files trimmed to contain just the type declarations, `resize` as a C external and `add_char`, on my machine I can observe a size reduction for the recursive version (.text is the code segment):

```
$ size --format=GNU add.o addrec.o
      text       data        bss      total filename
       199        320          0        519 add.o
       135        344          0        479 addrec.o
```

add.ml:

```ocaml
type inner_buffer = {
  buffer: bytes;
  length: int;
}

type t =
 {mutable inner : inner_buffer;
  mutable position : int;
  initial_buffer : bytes}

external resize : t -> int -> unit = "resize"

let add_char b c =
  let pos = b.position in
  let {buffer;length} = b.inner in
  if pos >= length then (
    resize b 1;
    Bytes.set b.inner.buffer pos c
  ) else
    Bytes.unsafe_set buffer pos c;
  b.position <- pos + 1
```

addrec.ml:

```
type inner_buffer = {
  buffer: bytes;
  length: int;
}
type t =
 {mutable inner : inner_buffer;
  mutable position : int;
  initial_buffer : bytes}

external resize : t -> int -> unit = "resize"

let rec add_char b c =
  let pos = b.position in
  let {buffer;length} = b.inner in
  if pos >= length then (
    resize b 1;
    add_char b c
  ) else (
    Bytes.unsafe_set buffer pos c;
    b.position <- pos + 1;
  )
```

Reply:
You are right,  comparing the code segment shows that there is a small decrease in size for the rec version:

```
      text       data        bss      total filename
      8554       6648          0      15202 indirection.o
      8490       6744          0      15234 indirection_rec.o
      8670       6752          0      15422 indirection_resize_once.o
      8542       6840          0      15382 indirection_resize_once_spill.o
```